spring-security-oauth2/src/main/java/org/springframework/security/oauth2/common/util/SerializationStrategy.java

```diff
@@ -0,0 +1,99 @@
+package org.springframework.security.oauth2.common.util;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.NotSerializableException;
+import java.io.ObjectInputStream;
+import java.io.ObjectStreamClass;
+import java.util.Collections;
+import java.util.List;
+import org.springframework.util.ClassUtils;
+
+/**
+ * Defines how objects are serialized an deserialized.
+ *
+ * @author Artem Smotrakov
+ */
+public interface SerializationStrategy {
+
+  /**
+   * Serializes an object.
+   *
+   * @param object The object to be serialized.
+   * @return A byte array.
+   */
+  byte[] serialize(Object object);
+
+  /**
+   * Deserializes an object from a byte array.
+   *
+   * @param byteArray The byte array.
+   * @param <T> The type of the object.
+   * @return The deserialized object.
+   */
+  <T> T deserialize(byte[] byteArray);
+
+  /**
+   * Special ObjectInputStream subclass that checks if classes are allowed to deserialize.
+   * The class should be configured with a whitelist of only allowed (safe) classes to deserialize.
+   */
+  class WhitelistedObjectInputStream extends ObjectInputStream {
```

Comment:
Let's move this to `WhitelistedSerializationStrategy` as a `private class` since it's part of that implementation.

Reply:
Okay, no problem.